heartbeat/monitors/active/dialchain/tlsmeta/tlsmeta.go

```diff
@@ -0,0 +1,133 @@
+	"crypto/x509"
+	"fmt"
+	"time"
+
+	"github.com/elastic/beats/v7/heartbeat/look"
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/common/transport/tlscommon"
+)
+
+func AddTLSMetadata(fields common.MapStr, connState cryptoTLS.ConnectionState, duration time.Duration) {
+	fields.Put("tls.established", true)
+	fields.Put("tls.rtt.handshake", look.RTT(duration))
+	versionDetails := tlscommon.TLSVersion(connState.Version).Details()
+	// The only situation in which versionDetails would be nil is if an unknown TLS version were to be
+	// encountered. Not filling the fields here makes sense, since there's no standard 'unknown' value.
+	if versionDetails != nil {
+		fields.Put("tls.version_protocol", versionDetails.Protocol)
+		fields.Put("tls.version", versionDetails.Version)
+	}
+
+	if connState.NegotiatedProtocol != "" {
+		fields.Put("tls.next_protocol", connState.NegotiatedProtocol)
+	}
+	fields.Put("tls.cipher", tlscommon.ResolveCipherSuite(connState.CipherSuite))
+
+	AddCertMetadata(fields, connState.PeerCertificates)
+}
+
+func AddCertMetadata(fields common.MapStr, certs []*x509.Certificate) {
+	hostCert := certs[0]
+
+	x509Fields := common.MapStr{}
+	serverFields := common.MapStr{"x509": x509Fields}
+	tlsFields := common.MapStr{"server": serverFields}
+
+	serverFields.Put("hash.sha1", fmt.Sprintf("%x", sha1.Sum(hostCert.Raw)))
+	serverFields.Put("hash.sha256", fmt.Sprintf("%x", sha256.Sum256(hostCert.Raw)))
+
+	x509Fields.Put("issuer.common_name", hostCert.Issuer.CommonName)
+	x509Fields.Put("issuer.distinguished_name", hostCert.Issuer.String())
+	x509Fields.Put("subject.common_name", hostCert.Subject.CommonName)
+	x509Fields.Put("subject.distinguished_name", hostCert.Subject.String())
+	x509Fields.Put("serial_number", hostCert.SerialNumber.String())
+	x509Fields.Put("signature_algorithm", hostCert.SignatureAlgorithm.String())
+	x509Fields.Put("public_key_algorithm", hostCert.PublicKeyAlgorithm.String())
+	if rsaKey, ok := hostCert.PublicKey.(*rsa.PublicKey); ok {
+		sizeInBits := rsaKey.Size() * 8
+		x509Fields.Put("public_key_size", sizeInBits)
+		x509Fields.Put("public_key_exponent", rsaKey.E)
+	} else if dsaKey, ok := hostCert.PublicKey.(*dsa2.PublicKey); ok {
+		if dsaKey.Parameters.P != nil {
+			x509Fields.Put("public_key_size", len(dsaKey.P.Bytes())*8)
+		} else {
+			x509Fields.Put("public_key_size", len(dsaKey.P.Bytes())*8)
+		}
+	} else if ecdsa, ok := hostCert.PublicKey.(*ecdsa.PublicKey); ok {
+		x509Fields.Put("public_key_curve", ecdsa.Curve.Params().Name)
+	}
+
+	chainNotBefore, chainNotAfter := calculateCertTimestamps(certs)
+	// Legacy non-ECS field
+	tlsFields.Put("certificate_not_valid_before", chainNotBefore)
+	x509Fields.Put("not_before", chainNotBefore)
+	if chainNotAfter != nil {
+		// Legacy non-ECS field
+		tlsFields.Put("certificate_not_valid_after", *chainNotAfter)
+		x509Fields.Put("not_after", *chainNotAfter)
+	}
+
+	fields.DeepUpdate(common.MapStr{"tls": tlsFields})
+}
```

Comment:
It might be easier to test and use less memory to create a `TLSMeta` struct and place that inside of the event.  Using `map[string]interface{}` everywhere is not very memory efficent.  Example is #16538 where it was converted to a struct.

Its not really a blocker, just something that could help.

Reply:
Discussed this with @simitt . Looks like this is an area @urso is working on, but isn't possible today because beats processors expect all event fields to be maps. I agree that it would be nice.